File: PHP/Smith_Bryan_Bankreport.php
<?php
///////////////////////////////////////////////////////////////////////////////
// CSC 4200-01 - Programming Languages
// Name: Bryan Smith
// Date: 3/5/14
// Description: Print bank statement based on data read from
//              file
// Parameters:
//   fileLoc - Path to the file to read.
///////////////////////////////////////////////////////////////////////////////

//Disable warnings. Might want to remove for debugging.
error_reporting(E_ERROR | E_PARSE); 

//Get arguments from either command line or URL
$fileLoc=null;
if ($argv[1] != "" && $argv[1] != null) {
	$fileLoc=$argv[1];
} else {
	$fileLoc = htmlspecialchars($_GET["fileName"]);
	if ($fileLoc == "" || $fileLoc == null) $fileLoc="TestData.txt";
}

//Create an account object, load in the file and parse it. 
//Then output the statement.
$account = new Account(0,0.0,10.0);
$account->parseToAccount(readDataFile($fileLoc));
echo $account;

///////////////////////////////////////////////////////////////////////////////

//Reads in a file returning the array of lines of what was read.
function readDataFile($fileLoc){
    try {
        $contents = array();
        $file = fopen( $fileLoc, "r" );
        if( $file == false )
        {
			echo "\nError in opening file ".$fileLoc."  Does it exist?\n\n";
			if (php_sapi_name() != 'cli') echo "You can also pass in the file name by adding ?fileName=FILENAME.txt  after the .php in the URL";
			else echo "You can also pass in the file name by adding it as a command line argument";
			exit();
        } else {
            while (($line = fgets($file)) !== false) {
                array_push($contents,$line);
            }   
        }
        fclose( $file );
        return $contents;
    } catch (Exception $e) {
        echo "\nError in opening file ".$fileLoc."  Does it exist?\n\n".$e->getMessage();
    }
}

//****************
//  Account Class
//****************
class Account {
	private $id = 0;
    private $balance = 0.0;
    private $initialBalance = 0.0;
    private $transactions = array();
    private $numbOfOverdrafts = 0;
    private $overdraftedLast = false;
    private $overdraftCharge = 10.0;
    
    public function Account($i,$b,$o=10.0) {
        $this->id = $i;
        $this->balance = $b;
        $this->initialBalance = $b;
        $this->overdraftCharge = $o;
    }
    
	//Parses data from an array into data to load into the account
    public function parseToAccount($s){
        try {
            $idString = explode(":",$s[0]);
            if ($this->startsWith($idString[0],"a",false)){
                $this->id = trim($idString[1]);
            } else {throw new Exception("Could not parse account number.");}
            $iBString = explode(":",$s[1]);
            if ($this->startsWith($iBString[0],"b",false)){
                $this->initialBalance = trim($iBString[1]);
                $this->balance = $this->initialBalance;
            } else {throw new Exception("Could not parse initial balance.");}

            for ($i=2;$i<count($s);$i++){
                $splitString = explode(":",$s[$i]);
                if ($this->startsWith($splitString[0],"d",false)) 
                    $this->addTransaction(new Deposit(trim($splitString[1]),trim($splitString[2])));
                else if ($this->startsWith($splitString[0],"w",false))
                    $this->addTransaction(new Withdrawal(trim($splitString[1]),trim($splitString[2])));
                else
                  throw new Exception("Failed to parse transactions.\n");  
            }
            $this->calculateBalance();
        } catch (Exception $e) {
            echo "\nError in parsing data.\n\n".$e->getMessage();
        } 
    }
    
	//Adds a transaction to the account
    private function addTransaction($t){
        try {
            array_push($this->transactions,$t);
        } catch (Exception $e) {
            echo "\nError adding transaction.\n\n".$e->getMessage();
        } 
    }
    
	 //Calculates the final balance and any overdrafts
    public function calculateBalance(){
        $this->balance = $this->initialBalance;
        foreach ($this->transactions as $t)
        {
            if ($t instanceof Deposit) {
                $this->balance += $t->getAmount();
            } else {
                $this->balance -= $t->getAmount();
            }
            
            if ($this->balance < 0 && !$this->overdraftedLast){
                $this->numbOfOverdrafts++;
                $this->overdraftedLast=true;
                $this->balance -= $this->overdraftCharge;
            } else if ($this->balance >= 0 && $this->overdraftedLast) {
                $this->overdraftedLast=false; 
            }
        }
    }
    
	//Determine if a sting starts with a specified character
    private function startsWith($haystack,$needle,$case=true){
        if($case)
            return strpos(trim($haystack), $needle, 0) === 0;
        return stripos(trim($haystack), $needle, 0) === 0;
    }
    
	//Print withdrawal summary
    private function printWithdrawals(){
        $totalWithdrawalAmount = 0.0;
        $tmpBalance = $this->initialBalance;
        $isFirstListed=false;
        
        echo "Summary of Withdrawals:	 amount	        Running Balance\n";
        echo "-------------------------------------------------------\n";
        foreach ($this->transactions as $t) {
            if ($t instanceof Deposit) { 
                echo "--deposit(see below)";
                $tmpBalance+=$t->getAmount();
            } else {
                $t->printTrans(false,!$isFirstListed); 
                $isFirstListed = true;
                $totalWithdrawalAmount += $t->getAmount();
                if ($tmpBalance <0 ) $this->overdraftedLast=true; 
                else $this->overdraftedLast=false;
                $tmpBalance -= $t->getAmount();
                if ($tmpBalance < 0 && $this->overdraftedLast == false) {
                    echo sprintf("%26s",sprintf("( %.2f)*",-$tmpBalance));   
                } else {
                    echo sprintf("%24s",sprintf("%.2f",$tmpBalance));  
                }
            }
            echo "\n";
        }
        echo "-------------------------------------------------------\n";
        echo sprintf("Total Withdrawals %13s\n",sprintf("$%.2f",$totalWithdrawalAmount));
    }
    
	//Print deposit summary
    private function printDeposits(){
        $totalDepositAmount = 0.0;
        $isFirstListed=false;

        echo "\nSummary of Deposits:\n";
        echo "-------------------------------\n";
        foreach ($this->transactions as $t) {
            if ($t instanceof Deposit) { 
                $t->printTrans(true,!$isFirstListed); 
                $isFirstListed = true;
                $totalDepositAmount += $t->getAmount();
            }
        }
        echo "-------------------------------\n";
        echo sprintf("Total Deposits %16s\n",sprintf("$%.2f",$totalDepositAmount));
    }
    
	//Print whole account statement
    public function __toString() {
		if (php_sapi_name() != 'cli') 
			echo "<html><body><pre style='word-wrap: break-word; white-space: pre-wrap;'>";
        echo "Statement for Account: ".$this->id."\n\n";
        echo "Beginning Balance: $".$this->initialBalance."\n\n";
        $this->printWithdrawals();
        $this->printDeposits();
        
        if ($this->numbOfOverdrafts>0){
            echo "\n*Overdraft Fees\n";
            echo "--------------\n";
            for ($i = 0; $i < $this->numbOfOverdrafts;$i++){
                echo sprintf("$%.2f\n",$this->overdraftCharge);
            }
            echo "--------------\n";
        }
        if ($this->balance <0)
            echo sprintf("\nEnding Balance: %15s\n",sprintf("( %.2f)",-$this->balance));
        else
            echo sprintf("\nEnding Balance: %15s\n",sprintf("$%.2f",$this->balance));
		
		if (php_sapi_name() != 'cli') echo "</pre></body></html>";
        return "";
    }
    
}

//****************
//  Transaction Class
//****************
class Transaction {
	private $source = "Default Source";
	private $amount = 0.0;
    
    public function Transaction($s,$a) {
        $this->source = $s;
        $this->amount = $a;
    }
	
	public function setSource($s){
		$this->source = $s;
	}
	
	public function getSource(){
		return $this->source;
	}
	
	public function setAmount($a){
		$this->amount = $a;
	}
	
	public function getAmount(){
		return $this->amount;
	}
	
	//Print the transaction
	public function printTrans($newLine, $cur) {
		if ($cur)
            echo sprintf("%-21s %9s",$this->getSource(),sprintf("$%.2f",$this->getAmount()));
        else
            echo sprintf("%-21s %9.2f",$this->getSource(),$this->getAmount());
        
        if ($newLine) echo "\n";

	}
}

//****************************
//  Withdrawal and Deposit Classes
//****************************
class Withdrawal extends Transaction{}
class Deposit extends Transaction {}
?> 

File: Ruby/Smith_Bryan_Bankreport.rb
#!/usr/bin/ruby
############################################
## CSC 4200-01 - Programming Languages
## Name: Bryan Smith
## Date: 3/21/14
## Description: Print bank statement based on data read from file
############################################


##################################
#   Transaction Classes
##################################
class Transaction

    def initialize(source, amount)
        @source = source
        @amount = amount
    end
    
    def getSource();@source;end
    def getAmount();@amount.to_f;end
    
    def print(newline,cur)
		if (cur)
            printf("%-21s %9s",getSource(),sprintf("$%.2f",getAmount()))
        else
            printf("%-21s %9.2f",getSource(),getAmount())
        end
        
        if (newline);puts "\n";end
    end
            
end

class Withdrawal < Transaction;end  #Withdrawal class
class Deposit < Transaction;end     #Deposit class

##################################
#   Account Class
##################################
class Account
    def initialize(i, iB,odc)
        @id=i
        @balance=iB.to_f
        @initialBalance=iB.to_f
        @overdraftCharge=odc.to_f
        @transactions=Array.new
        @numbOfOverdrafts=0
        @overdraftedLast=false
    end
    
    def getTransactions();@transactions;end;
    
    def calculateBalance()
        @balance = @initialBalance.to_f
        @transactions.each do |t|
            if (t.instance_of?(Deposit))
                @balance+=t.getAmount().to_f
            else
                @balance-=t.getAmount().to_f
            end
            if (@balance<0 and !@overdraftedLast)
                @numbOfOverdrafts+=1
                @overdraftedLast=true
            end
            if (@balance >=0 and @overdraftedLast); @overdraftedLast=false; end;
        end
    end
    
    def addTransaction(t);@transactions << t;end;
    
    def parseToAccount(s)
        begin
            @idString=s[0].split(':')
            if(@idString[0].start_with?("acc"));@id = @idString[1].strip();
            else; raise "Cannot parse Account number.";end;
            @iBString=s[1].split(':')
            if(@iBString[0].start_with?("beginning"))
                @initialBalance = @iBString[1].strip()
                @balanace = @initialBalance
            else; raise "Cannot parse initial balance.";end;
            (2..s.size-1).each do |i|
                splitString = s[i].split(':')
                if (splitString[0].to_s[0].downcase! == 'd')
                    addTransaction(Deposit.new(splitString[1].strip(),splitString[2].strip()))
                elsif (splitString[0].to_s[0].downcase! == 'w')
                    addTransaction(Withdrawal.new(splitString[1].strip(),splitString[2].strip()))
                else
                    raise "Could not parse transactions."
                end
            end
            calculateBalance()
            self
        rescue
            raise "Error parsing data."
            abort
        end
    end

    def printWithdrawals()
        @totalWithdrawalAmount = 0.0
        @tmpBalance = @initialBalance.to_f
        @isFirstListed=false
        puts "Summary of Withdrawals:  amount	        Running Balance\n"
        puts "-------------------------------------------------------\n"
        @transactions.each do |t|
            if (t.instance_of?(Deposit))
                puts "--deposit(see below)\n"
                @tmpBalance+=t.getAmount().to_f
            elsif (t.instance_of?(Withdrawal))
                t.print(false,!@isFirstListed)
                @isFirstListed=true
                @totalWithdrawalAmount+= t.getAmount().to_f
                if(@tmpBalance <0.0);@overdraftedLast=true;
                else; @overdraftedLast=false;end;
                @tmpBalance-=t.getAmount().to_f
                if(@tmpBalance<0 and !@overdraftedLast)
                    printf("%26s\n",sprintf("( %.2f)*",-@tmpBalance));
                else
                    printf("%24s\n",sprintf("%.2f",@tmpBalance));
                end
            end

        end
         puts "-------------------------------------------------------\n"
         printf("Total Withdrawals %13s\n",sprintf("$%.2f",@totalWithdrawalAmount))
    end

    def printDeposits()
        @totalDepositAmount = 0.0
        @isFirstListed=false
        puts "\nSummary of Deposits:\n"
        puts "-------------------------------\n"
        @transactions.each do |t|
            if (t.instance_of?(Deposit))
                t.print(true,!@isFirstListed)
                @isFirstListed=true
                @totalDepositAmount+= t.getAmount().to_f
            end
        end
        puts "-------------------------------\n"
        printf("Total Deposits %16s\n",sprintf("$%.2f",@totalDepositAmount))
    end

    def to_s
        print("Statement for Account: #@id\n\n")
        print("Beginning Balance: $#@initialBalance\n\n")
        printWithdrawals()
        printDeposits()
        if (@numbOfOverdrafts>0)
            print("\n*Overdraft Fees\n")
            print("--------------\n")
            i=0
            while i < @numbOfOverdrafts do
                printf("$%.2f\n",@overdraftCharge)
                @balance-=@overdraftCharge
                i=i+1
            end
            print("--------------\n")
        end
        if (@balance <0)
            s=sprintf("\nEnding Balance: %15s\n",sprintf("( %.2f)",-@balance));
        else
            s=sprintf("\nEnding Balance: %15s\n",sprintf("$%.2f",@balance));
        end
        s
    end

end

#Read in the account data file
def readFile(fileLoc)
    begin
        arr = IO.readlines(fileLoc);arr
    rescue
        puts "Cannot open account file. Does it exist?"
        abort
    end
end

##################################
#   Execute
##################################
@account = Account.new(123,0.0,10.0)
puts @account.parseToAccount(readFile("account.txt"))
